code/modules/pool/pool_controller.dm

```diff
@@ -0,0 +1,330 @@
+	else
+		to_chat(user, "<span class='warning'>The interface on [src] is already too damaged to short it again.</span>")
+		return
+
+/obj/machinery/pool/controller/attackby(obj/item/W, mob/user)
+	if(shocked && !(stat & NOPOWER))
+		shock(user,50)
+	if(stat & (BROKEN))
+		return
+
+	if(istype(W,/obj/item/reagent_containers))
+		if(W.reagents.total_volume >= 100) //check if there's enough reagent
+			for(var/datum/reagent/R in W.reagents.reagent_list)
+				if(R.id in GLOB.blacklisted_pool_reagents)
+					to_chat(user, "\The [src] cannot accept [R.name].")
+					reagents.clear_reagents()
+					return
+				if(R.reagent_state == SOLID)
+					to_chat(user, "The pool cannot accept reagents in solid form!.")
+					reagents.clear_reagents()
+					return
+			reagents.clear_reagents()
+			W.reagents.copy_to(reagents, 100)
+			W.reagents.clear_reagents()
+			user.visible_message("<span class='notice'>\The [src] makes a slurping noise.</span>", "<span class='notice'>All of the contents of \the [W] are quickly suctioned out by the machine!</span")
+			updateUsrDialog()
+			var/reagent_names = ""
+			for(var/datum/reagent/R in reagents.reagent_list)
+				reagent_names += "[R.name], "
+			log_game("[key_name(user)] has changed the [src] chems to [reagent_names]")
+			message_admins("[key_name_admin(user)] has changed the [src] chems to [reagent_names].")
+			interact_delay = world.time + 15
+		else
+			to_chat(user, "<span class='notice'>\The [src] beeps unpleasantly as it rejects the beaker. It must not have enough in it.</span>")
+			return
+	else if(panel_open && is_wire_tool(W))
+		wires.interact(user)
+	else
+		return ..()
+
+/obj/machinery/pool/controller/screwdriver_act(mob/living/user, obj/item/W)
+	. = ..()
+	if(.)
+		return TRUE
+	cut_overlays()
+	panel_open = !panel_open
+	to_chat(user, "You [panel_open ? "open" : "close"] the maintenance panel.")
+	W.play_tool_sound(src)
+	if(panel_open)
+		add_overlay("wires")
+	return TRUE
+
+//procs
+/obj/machinery/pool/controller/proc/shock(mob/user, prb)
+	if(stat & (BROKEN|NOPOWER))		// unpowered, no shock
+		return FALSE
+	if(!prob(prb))
+		return FALSE
+	var/datum/effect_system/spark_spread/s = new /datum/effect_system/spark_spread
+	s.set_up(5, 1, src)
+	s.start()
+	if(electrocute_mob(user, get_area(src), src, 0.7))
+		return TRUE
+	else
+		return FALSE
+
+/obj/machinery/pool/controller/proc/poolreagent()
+	if(reagents.reagent_list.len > 0)
+		for(var/turf/open/pool/W in linkedturfs)
+			for(var/mob/living/carbon/human/swimee in W)
+				for(var/datum/reagent/R in reagents.reagent_list)
+					if(R.reagent_state == SOLID)
+						R.reagent_state = LIQUID
+					swimee.reagents.add_reagent(R.id, 0.5) //osmosis
+				reagents.reaction(swimee, VAPOR, 0.03) //3 percent
+			for(var/obj/objects in W)
+				if(W.reagents)
+					W.reagents.reaction(objects, VAPOR, 1)
+	reagent_delay = world.time + POOL_REAGENT_TICK_INTERVAL
+	changecolor()
+
+
+/obj/machinery/pool/controller/process()
+	updateUsrDialog()
```

Comment:
these should happen at the end of processing

Reply:
I still need to rework the ui, it's missing a bunch of functionality I had to remove due to changes in how the reagent changing works.